### DISCUSSION TITLE
Change uibRoot directory

### DISCUSSION THREAD

**Discussion by anxuae** [TotallyInformation/node-red-contrib-uibuilder#138] [Q&A] [answered]
Hello,

Is it possible to change the `uibRoot` directory to a custom sub-directories tree?

I'm using the project mode and I would like to put the `uibuilder` folder in a sub-directory. It could be nice if I can configure it in the `package.json` file. For instance:

```json
    "node-red": {
        "settings": {
            "flowFile": "flow.json",
            "credentialsFile": "flow_cred.json",
            "uibRoot": "path/to/uibuilder"
        }
    }
```

**Comment by TotallyInformation** [comment, marked as answer]
Hi Anxuae, I'm afraid that isn't possible right now though it is on the backlog.

As the last release reintroduced some additional settings, the way is open again to add that as an option and ~~I'll add~~ added to the [kanban board](https://github.com/TotallyInformation/node-red-contrib-uibuilder/projects/1) so I don't forget.

Of course, if you wanted to contribute a change to the code, I would certainly be open to it 😁

**Reply by anxuae** [reply]
Hi @TotallyInformation,
Thanks for your quick answer, I'm afraid to be a newbie in js 😄 , but I could have a look

**Reply by TotallyInformation** [reply]
No worries, always worth asking. Feel free to look at the code, it is a good way to learn 😀

In need to do some work on uibuilder anyway so I'll see if I can squeeze this into the next release.

In the meantime, if you want to maintain the uibRoot somewhere else, you should try using a soft-link from your chosen location to the uibuilder standard location. Exactly how you do that is dependent on your operating system. But basically, you shut down Node-RED, move the uibuilder root folder to wherever you want it then create a soft-link back to its original location.

**Reply by TotallyInformation** [reply]
Incidentally, the settings.js will look something like this once the work is done:

```js
    /** Custom settings for all uibuilder node instances */
    uibuilder: {
        /** Optional HTTP PORT. 
         * If set and different to Node-RED's uiPort, uibuilder will create
         * a separate webserver for its own use.
         */
        //port: process.env.UIBPORT || 3001,
       /** Optional: Change location of uibRoot
        * If set, instead of something like `~/.node-red/uibuilder`, the uibRoot folder can be anywhere you like.
        */
       //uibRoot: process.env.UIBROOT || '/uibRoot', //path.join(os.homedir(), 'myuibroot')',
    },
```

**Reply by anxuae** [reply]
Hi, thanks for it!
Because I'm using the "project" mode in order to put my projects on GitLab, is not possible set soft link: as you say, it is OS-dependent and can not commit it on git (will not work for all users).

Is it possible to define the `uibRoot` per project?
(or define it depending on the project path)

**Reply by anxuae** [reply]
Hi @TotallyInformation , how to define the `uibRoot` according to the project path?
Something like that could work:
```
uibRoot: path.join(process.env.PROJECT, 'myuibroot'),
```

**Reply by TotallyInformation** [reply]
Hi again. v4 is now in beta and available by manually installing from GitHub with `npm install totallyinformation/node-red-contrib-uibuilder`. It includes the ability to set the uibRoot folder from settings.js as indicated above.

Please note that there is a bug in the current beta that means it doesn't play nicely with Node.js v12 and instead needs v14. I'll be fixing that shortly and then I expect to be able to publish v4 to npm soon, hopefully the weekend.
